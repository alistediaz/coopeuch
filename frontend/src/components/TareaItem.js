import React, { useRef, useState } from "react";
import { deleteTarea, updateTarea } from "../redux/tareaSlice";
import { selectTareaById } from "../redux/tareaSlice";
import { useDispatch, useSelector } from "react-redux";
import {
  AiOutlineForm,
  AiFillSave,
  AiOutlineCarryOut,
  AiOutlineDelete,
  AiOutlineStop,
} from "react-icons/ai";

export const TareaItem = (props) => {
  const { item } = props;
  const [isEditing, setIsEditing] = useState(false);
  const inputRef = useRef(true);
  const tarea = useSelector((state) => selectTareaById(state, item.id));
  const dispatch = useDispatch();

  const changeFocus = () => {
    inputRef.current.disabled = false;
    inputRef.current.focus();
    setIsEditing(!isEditing);
  };

  const callUpdateTarea = async (tarea) => {
    await dispatch(updateTarea(tarea));
  };

  const update = (id, value) => {
    if (value === "") {
      alert("Debe ingresar descripciÃ³n de la tarea");
      inputRef.current.value = tarea.descripcion;
    } else {
      if (value !== tarea.descripcions) {
        callUpdateTarea({ id, descripcion: value, vigente: tarea.vigente });
      }
      setIsEditing(false);
      inputRef.current.disabled = true;
    }
  };

  const updateVigente = (item) => {
    callUpdateTarea({ ...item, vigente: !item.vigente });
  };

  const eliminaTarea = async (tareaId) => {
    await dispatch(deleteTarea(tareaId));
  };
  return (
    <li
      key={item.id}
      className="card"
      style={!item.vigente ? { backgroundColor: "Ivory" } : {}}
    >
      <textarea
        ref={inputRef}
        disabled={inputRef}
        defaultValue={item.descripcion}
        maxLength={100}
      />
      <div className="btns">
        {!isEditing && (
          <button onClick={() => changeFocus()} title="Editar">
            <AiOutlineForm />
          </button>
        )}
        {isEditing && (
          <button
            onClick={() => update(item.id, inputRef.current.value)}
            title="Grabar"
          >
            <AiFillSave />
          </button>
        )}
        {item.vigente && (
          <button onClick={() => updateVigente(item)} title="Vigente">
            <AiOutlineCarryOut />
          </button>
        )}
        {!item.vigente && (
          <button
            style={{ color: "grey" }}
            onClick={() => updateVigente(item)}
            title="No Vigente"
          >
            <AiOutlineStop />
          </button>
        )}

        <button style={{ color: "red" }} onClick={() => eliminaTarea(item.id)} title="Eliminar">
          <AiOutlineDelete />
        </button>
      </div>

      {item.completed && <span className="completed">done</span>}
    </li>
  );
};
