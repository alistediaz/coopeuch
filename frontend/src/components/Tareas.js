import React, { useState, useEffect } from "react";
import { addNewTarea } from "../redux/tareaSlice";
import { useDispatch, useSelector } from "react-redux";

export const Tareas = () => {
  const [tarea, setTarea] = useState("")
  const [vigente, setVigente] = useState(false)
  const [disabledButton, setDisabledButton] = useState(false)
  // @ts-ignore
  const tareaStatus = useSelector((state) => state.status)
  const dispatch = useDispatch()

  const add = async () => {
    if (tarea === "") {
      alert("Debe ingresar descripciÃ³n de la tarea");
    } else {
      setDisabledButton(true)
      await dispatch(
        addNewTarea({
          descripcion: tarea,
          vigente: vigente,
        })
      ).unwrap();
      setTarea("");
      setVigente(false);
    }
  };

  useEffect(() => {
    if (tareaStatus === "idle" || tareaStatus === "succeeded") {
      setDisabledButton(false)
    }
  }, [tareaStatus, dispatch]);

  const handleChange = (e) => {
    setTarea(e.target.value);
  };

  const handleChangeVigente = (e) => {
    setVigente(e.target.checked);
  };

  return (
    <div className="addTareas">
      <input
        type="text"
        onChange={(e) => handleChange(e)}
        className="tarea-input"
        value={tarea}
        maxLength={100}
        title="DescripciÃ³n"
      />
      <div className="vigente">
        Vigente:
        <input type="checkbox" onChange={(e) => handleChangeVigente(e)} />
      </div>
      <button className="add-btn" onClick={() => add()} disabled={disabledButton}>
        Nueva
      </button>
    </div>
  );
};
