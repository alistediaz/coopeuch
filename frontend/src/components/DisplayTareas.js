import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { selectAllTareas, fetchTareas } from "../redux/tareaSlice";
import { TareaItem } from "./TareaItem";

export const DisplayTareas = () => {
  const dispatch = useDispatch();
  // @ts-ignore
  const tareaStatus = useSelector((state) => state.status);
  // @ts-ignore
  const error = useSelector((state) => state.error);
  const tareas = useSelector(selectAllTareas);

  useEffect(() => {
    if (tareaStatus === "idle") {
      dispatch(fetchTareas());
    }
  }, [tareaStatus, dispatch]);

  let content;

  if (tareaStatus === "loading") {
    content = <h3>Leyendo...</h3>;
  } else if (tareaStatus === "succeeded") {
    content = tareas.map((tarea) => {
      return <TareaItem key={tarea.id} item={tarea} />;
    });
  }

  return (
    <div className="displaytareas">
      {tareaStatus === "failed" && <div>{error}</div>}
      <ul>{content}</ul>
    </div>
  );
};
