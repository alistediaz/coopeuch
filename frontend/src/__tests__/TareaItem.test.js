import React from "react";
import renderer from "react-test-renderer";
import { TareaItem } from "../components/TareaItem";

const tarea = {
  id: 33,
  descripcion: "test",
  fechaCreacion: "2023-07-10T22:05:09.882+00:00",
  vigente: true,
};

jest.mock("react-redux", () => {
  return {
    useDispatch: jest.fn(),
    useSelector: () => {
      return tarea;
    },
  };
});

describe("<TareaItem />", () => {
  const defaultProps = { key: 33, item: tarea };
  const wrapper = renderer.create(
      <TareaItem {...defaultProps} />
  );

  test("render TareaItem", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
