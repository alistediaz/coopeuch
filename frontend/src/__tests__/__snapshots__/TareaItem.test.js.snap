// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TareaItem /> render TareaItem 1`] = `
<li
  className="card"
  style={Object {}}
>
  <textarea
    defaultValue="test"
    disabled={
      Object {
        "current": null,
      }
    }
    maxLength={100}
  />
  <div
    className="btns"
  >
    <button
      onClick={[Function]}
      title="Editar"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M904 512h-56c-4.4 0-8 3.6-8 8v320H184V184h320c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V520c0-4.4-3.6-8-8-8z"
        />
        <path
          d="M355.9 534.9L354 653.8c-.1 8.9 7.1 16.2 16 16.2h.4l118-2.9c2-.1 4-.9 5.4-2.3l415.9-415c3.1-3.1 3.1-8.2 0-11.3L785.4 114.3c-1.6-1.6-3.6-2.3-5.7-2.3s-4.1.8-5.7 2.3l-415.8 415a8.3 8.3 0 0 0-2.3 5.6zm63.5 23.6L779.7 199l45.2 45.1-360.5 359.7-45.7 1.1.7-46.4z"
        />
      </svg>
    </button>
    <button
      onClick={[Function]}
      title="Vigente"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v584zM688 420h-55.2c-5.1 0-10 2.5-13 6.6L468.9 634.4l-64.7-89c-3-4.1-7.8-6.6-13-6.6H336c-6.5 0-10.3 7.4-6.5 12.7l126.4 174a16.1 16.1 0 0 0 26 0l212.6-292.7c3.8-5.4 0-12.8-6.5-12.8z"
        />
      </svg>
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "red",
        }
      }
      title="Eliminar"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
        />
      </svg>
    </button>
  </div>
</li>
`;
